src/control/ControlModule.cpp

```diff
@@ -0,0 +1,85 @@
+//
+// Created by jaro on 15-10-20.
+//
+
+#include "control/ControlModule.h"
+#include <roboteam_utils/Print.h>
+#include "control/ControlUtils.h"
+#include "utilities/IOManager.h"
+#include "utilities/Settings.h"
+#include "world/World.hpp"
+#include "iostream"
+
+namespace rtt::ai::control {
+
+
+    std::vector<proto::RobotCommand> ControlModule::robotCommands;
+
+    void ControlModule::rotateRobotCommand(proto::RobotCommand& command){
+        command.mutable_vel()->set_x(-command.vel().x());
+        command.mutable_vel()->set_y(-command.vel().y());
+        command.set_w(static_cast<float>(Angle(command.w() + M_PI)));
+    }
+
+    void ControlModule::limitRobotCommand(proto::RobotCommand& command,std::optional<rtt::world::view::RobotView> robot) {
+        limitVel(command,robot);
+        limitAngularVel(command,robot);
+    }
+
+    void ControlModule::limitVel(proto::RobotCommand& command,std::optional<rtt::world::view::RobotView> robot) {
+        auto limitedVel = Vector2(command.vel().x(), command.vel().y());
+
+        limitedVel = control::ControlUtils::velocityLimiter(limitedVel);
+
+        if (std::isnan(limitedVel.x) || std::isnan(limitedVel.y)) {
+            RTT_ERROR("A certain Skill has produced a NaN error. \nRobot: " + std::to_string(robot.value()->getId()))
+        }
+
+        // Limit robot velocity when the robot has the ball
+        if (robot->hasBall() && limitedVel.length() > stp::control_constants::MAX_VEL_WHEN_HAS_BALL) {
+            // Clamp velocity
+            limitedVel = control::ControlUtils::velocityLimiter(limitedVel, stp::control_constants::MAX_VEL_WHEN_HAS_BALL,
+                                                                0.0, false);
+        }
+
+        command.mutable_vel()->set_x(static_cast<float>(limitedVel.x));
+        command.mutable_vel()->set_y(static_cast<float>(limitedVel.y));
```

Comment:
i wonder why we're using double, when we cast it to float before sending :P

Reply:
I mean, it doesn't matter much. Computing things in double is usually faster though, unless you're sending big matrices or large amount of numeric data where the memory overhead wins.